job/watcher.go

```diff
@@ -33,72 +34,105 @@ func NewWatcher(client *kubernetes.Clientset, container string) *Watcher {
+// So we must trace all Pods even though the Pod is failed.
+// And it isn't necessary to stop the loop because the Job is watched in WaitJobComplete.
 func (w *Watcher) Watch(job *v1.Job, ctx context.Context) error {
-	podList, err := w.WaitToStartPods(job)
-	if err != nil {
-		return err
+	currentPodList := []corev1.Pod{}
+retry:
+	for {
+		newPodList, err := w.FindPods(job)
+		if err != nil {
+			return err
+		}
+
+		incrementalPodList := diffPods(currentPodList, newPodList)
+		go w.WatchPods(incrementalPodList, ctx)
+
+		time.Sleep(1 * time.Second)
+		currentPodList = newPodList
+		continue retry
 	}
+	return nil
+}
 
+// WatchPods gets wait to start pod and tail the logs.
+func (w *Watcher) WatchPods(pods []corev1.Pod, ctx context.Context) error {
 	var wg sync.WaitGroup
-	errCh := make(chan error, len(podList.Items))
-	for _, pod := range podList.Items {
-		// Ref: https://github.com/kubernetes/client-go/blob/03bfb9bdcfe5482795b999f39ca3ed9ad42ce5bb/kubernetes/typed/core/v1/pod_expansion.go
-		logOptions := corev1.PodLogOptions{
-			Container: w.Container,
-			Follow:    true,
-		}
-		// Ref: https://stackoverflow.com/questions/32983228/kubernetes-go-client-api-for-log-of-a-particular-pod
-		request := w.client.CoreV1().Pods(pod.Namespace).GetLogs(pod.Name, &logOptions).Context(ctx).
-			Param("follow", strconv.FormatBool(true)).
-			Param("container", w.Container).
-			Param("timestamps", strconv.FormatBool(false))
+	errCh := make(chan error, len(pods))
+
+	for _, pod := range pods {
 		wg.Add(1)
 		go func() {
 			defer wg.Done()
-			err := readStreamLog(request, pod)
+			pod, err := w.WaitToStartPod(pod)
+			if err != nil {
+				errCh <- err
+				return
+			}
+			// Ref: https://github.com/kubernetes/client-go/blob/03bfb9bdcfe5482795b999f39ca3ed9ad42ce5bb/kubernetes/typed/core/v1/pod_expansion.go
+			logOptions := corev1.PodLogOptions{
+				Container: w.Container,
+				Follow:    true,
+			}
+			// Ref: https://stackoverflow.com/questions/32983228/kubernetes-go-client-api-for-log-of-a-particular-pod
+			request := w.client.CoreV1().Pods(pod.Namespace).GetLogs(pod.Name, &logOptions).Context(ctx).
+				Param("follow", strconv.FormatBool(true)).
+				Param("container", w.Container).
+				Param("timestamps", strconv.FormatBool(false))
+			err = readStreamLog(request, pod)
 			errCh <- err
 		}()
 	}
+
 	select {
 	case err := <-errCh:
 		if err != nil {
+			log.Error(err)
 			return err
 		}
 	}
 	wg.Wait()
 	return nil
 }
 
-// WaitToStartPods wait until starting the pods.
+// FindPods finds pods in
+func (w *Watcher) FindPods(job *v1.Job) ([]corev1.Pod, error) {
+	labels := parseLabels(job.Spec.Template.Labels)
+	listOptions := metav1.ListOptions{
+		LabelSelector: labels,
+	}
+	podList, err := w.client.CoreV1().Pods(job.Namespace).List(listOptions)
+	if err != nil {
+		return []corev1.Pod{}, err
+	}
+	return podList.Items, err
+}
+
+// WaitToStartPods wait until starting the pod.
```

Comment:
__[Golint]__ comment on exported method Watcher.WaitToStartPod should be of the form "WaitToStartPod ..." ([link](https://sider.review/gh/repos/169710725/pulls/15#issue-3828089))

<sub>posted by Sider</sub>